### PR TITLE
fix(auto-review): 保留真实授权与澄清限制，准确回传拒绝原因

### PR DESCRIPTION
**Pull request by dashhuang**
### 摘要

修复 Auto 审批拿不到真实用户授权、把默认工作目录当成绝对授权范围，以及拒绝原因被统一显示为用户拒绝的问题。审批依据取自 Host 接受的原始输入，助手交接文字和插件改写不能替代授权；规则放入模型 system 通道，待审证据独立传入。

恢复授权时包含已接受的澄清回答、批准计划及修改反馈，按接受时间排序后截取历史。超长输入保留原文到最终预算处理，无法完整保留限制时清掉此前授权。复用现有消息 metadata 和持久化队列，无 schema migration、独立缓存或新增重试。

### 变更类型

- [ ] `feat` 新功能
- [x] `fix` 缺陷修复
- [ ] `refactor` / `perf` 重构或性能优化
- [ ] `docs` / `test` / `chore` 文档、测试或工程维护
- [ ] 其他：

### 范围

- 关联 Issue / 需求：Auto 审批误拒调查及三个 harness 对称核查，承接 #3993。
- 本 PR 包含：真实输入捕获、历史授权恢复、模型请求与脱敏诊断、Claude Code/Pi/Cindy Host 工具的拒绝原因和取消归因。
- 明确不包含：Codex 原生源码/协议补丁、插话或新用户消息反馈、自动重试、Bot 触发和账号绑定变化。
- 用户可见变化：已授权的项目 worktree 操作可按真实范围接受审批；支持的工具在原结果中显示拒绝原因，取消不再假称人工拒绝。
- 是否存在 breaking change：无。Ask/Full Access 边界保留。Codex 官方原生 exec/patch/权限申请继续返回官方通用拒绝结果，具体理由仅在 Cindy 控制的动态工具结果中回传。

### UI 变化

- 引用的设计规范：不涉及视觉 UI；仅调整模型可见工具错误结果。

### 自动验证

- 通过 git skill `run-unit-gate.sh` 执行根 `pnpm test:unit`：GATE_EXIT=0，全量 workspace 单测通过。
- 根 `pnpm test:unit:related`：通过（Desktop、lizi-mcps、maker-core、orca-workflow）。
- Desktop `typecheck`：通过；maker-core `run --if-present typecheck`：无 script，跳过。
- 定向输入恢复/发送事务/卡片持久化/真实 SQLite 回归：20 + 89 + 124 + 43 项通过。
- `git diff --check`：通过；完整 diff 复审及独立授权/三 harness 复审：未发现剩余 P0/P1。

覆盖：真实输入盖章后超长撤销、后续消息保留问答/计划限制、迟到回答排序与截取、IPC 伪造与旧 renderer 展示回写、clear/rewind、附件变化、三个 harness 的拒绝与取消及迟到回调。

### 手工验证

已完成直接模型回放，未执行待审动作：最终 35 类案例各两次，共 70 次请求，67 次有效响应中 65 次符合预期；2 次缺少文件目标的动作被错误 allow，另有 2 次 HTTP400 和 1 次 HTTP503。最终这组有效请求均在 12 秒内完成；这些数字只代表该组回放，不是线上可靠性保证。事故对照中使用的重构授权是诊断假设，不当作历史原文。

参考 Hermes 的规则/证据分离和独立单次判断，未照搬其仅面向命令风险的范围：
https://github.com/NousResearch/hermes-agent/blob/22c5684b983eac6a81ee015ae80296c4b3dbf5bb/tools/approval_smart.py

### 未执行的验证

默认 Cindy 审批网关未做线上端到端验证；未验证 Windows/Linux 客户端和真实 SSH 环境。maker-core 没有 typecheck script，按 --if-present 跳过，未声称完整 tsc 通过。已撤回的 Codex 原生实验不作为本 PR 的交付证据。

### 风险分类

- [ ] 无已知风险
- [x] SQLite / migration
- [x] system prompt
- [x] 协议兼容
- [x] 权限 / 安全 / 用户数据
- [ ] 存量插件兼容（批准状态 / 指纹 / manifest 校验 / 安装布局 / 包格式）
- [ ] 原生层 / fingerprint / OTA
- [ ] 跨平台差异
- [ ] 其他：

### 影响与回滚

- 影响范围：三个 harness 的 Auto 授权上下文及审批拒绝反馈。独立审批 system prompt 修改已由维护者在本次需求中确认；不修改主 Agent system prompt，不增加每轮主模型网络调用或改变主模型缓存前缀。模型回放结果及延迟范围见上。
- 数据：复用 agentMeta.autoReviewUserText；问答/计划原始证据由 Host 与卡片内容原子写入，IPC 不能伪造或替换。旧记录缺少可信来源、顺序不可确认或上下文超预算时不恢复旧授权，可能要求重新明确范围。无数据库 migration。
- 兼容：Pi 保留旧 confirm 布尔返回兼容；Codex 原生协议不变。存量插件影响：无；未改 receipt、指纹、manifest、安装布局或包格式。共用插件文件交接仅修正取消提示；按插件基座规则，合并仍需指定放行人明确 Approve。
- 回滚 / 降级方式：回退本次代码即可，无安装包替换或数据迁移回滚。新增 metadata 不影响旧客户端读取消息；审批基础设施失败仍按既有流程转人工。

### 提交前检查

- [x] 已 review 完整 diff
- [x] 每个 commit 都带 DCO 签名（`git commit -s`）
- [x] UI 改动已在「UI 变化」注明引用的设计规范章节（不涉及 UI 则跳过）
- [x] 未提交凭证、令牌或授权文件
- [x] 已补充必要文档（本 PR 的行为与验证说明）
- [x] 已确认测试结果或说明未执行原因

### CODE REVIEW

**Review by zqchris — APPROVED** [review]

**Review by MagicLizi — APPROVED** [review]
## Code Review Report: PR #4078

**fix(auto-review): preserve real authorization and clarify restrictions, accurately return rejection reasons**

### Summary

This PR fixes three interrelated issues in the Auto-review permission system:
1. The reviewer was not receiving the user's actual authorized input (instead getting decorated/plugin-rewritten text)
2. The default working directory was treated as the absolute authorization scope
3. All denial reasons were displayed as generic "user denied" regardless of actual cause

### Review Scope

Standard-tier review of 1088-line diff across 35 files. Both maintainer signoff triggers (pluginBase and arch) were released by @zqchris via GitHub Approve. Security scan: pass (0 hard hits, 0 soft hits). CI: all green.

### Architecture Assessment

The changes establish a clear trust boundary for authorization evidence:

**IPC Protection (messages.ts)**: The renderer can no longer mint, replace, or remove `autoReviewUserText` metadata. The SQL-level CASE expressions preserve Host-authored evidence atomically across concurrent renderer card PATCHes. User content edits correctly invalidate authorization (role='user' only), while ask_user/plan_review interaction answers are independently preserved.

**Authorization Restoration (autoReviewUserIntent.ts)**: A new module restores bounded authorization history from persisted messages. Key design decisions are conservative and correct:
- Messages without Host-captured text (`autoReviewUserText`) reset the intent chain
- Interaction answers are ordered by acceptance time, not card creation time
- Millisecond ties between a card answer and another entry conservatively return empty intent
- Quotes, images, files, pasted ranges, and session references are not promoted to authorization
- Auto-resume, context-rebuild, and UI-trigger messages are excluded
- Oversized inputs that cannot be fully preserved invalidate preceding grants

**System/User Channel Separation (oneShotCandidates.ts)**: Review policy is placed on the native system channel; untrusted action/user evidence goes on the user channel. The split uses the `<review_input>` delimiter consistently across all model providers.

**Prompt Restructuring (auto-permission-reviewer.ts)**: The decision order is clarified (AUTHORITY -> SCOPE -> VERDICT). `writableRoots` renamed to `defaultWritableRoots` with clear semantics that these are standing grants, not complete scope. `authorizationContext` is always included (defaulting to owner/direct). Tool envelopes with `kind: 'other'` are now parsed as JSON when possible.

**Denial Reason Propagation**: `formatPermissionDenial` provides consistent, bounded (240-char) denial messages across all three harnesses (Claude Code, Codex, Pi). The Pi bridge's `permissionDenialReason` correctly decodes the extended `source:reason` format. Ghost.ts correctly distinguishes `session_closed`/`session_aborted` from user denial.

**Codex Approval Refactor**: The approval function signature changed from returning `ApprovalDecision` to `{decision, reason?}`. All 8 callers verified to correctly use `.decision`. The `settled` flag prevents late callbacks from overwriting system cancellation. The `denialReason` variable is properly scoped per approval call.

### Verified Items

- No credentials, tokens, or authorization files in the diff
- No database schema migrations; uses json functions on existing `agent_meta` column
- Pi retains old confirm boolean return compatibility
- Codex native exec/patch protocol unchanged; reasons only in Cindy-controlled dynamic tool results
- Plugin base change (ghost.ts) is minimal messaging improvement, no behavior change to ok/error status
- IPC create handler strips `autoReviewUserText` before insert transaction
- `readAutoReviewUserText` returns null for any content with quotes, images, files, or references
- `restoreAutoReviewUserIntent` handles clear/rewind boundaries via database query filtering
- `redactSensitiveText` applied to model-returned reasons before logging

### Findings

**P0**: 0
**P1**: 0

No blocking issues found. The implementation is security-conscious with proper trust boundaries, conservative fallbacks, bounded text limits, and extensive test coverage across all boundary conditions.

### Bot Settle Notice

copilot-pull-request-reviewer and chatgpt-codex-connector did not post review comments within the 20-minute settle window (waited 271 minutes). Their opinions may arrive after merge. Codex security review completed with no findings. Greptile rated confidence 5/5 with no blocking issues.

File: apps/desktop/src/main/maker-ipc/autoReviewUserIntent.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  agentMeta: Record<string, unknown> | null;
+}
+
+function interactionAnswer(message: AutoReviewHistoryMessage): { text: string; acceptedAt: number } | null {
+  if (message.role !== 'ask_user' && message.role !== 'plan_review') return null;
+  const value = message.agentMeta?.autoReviewUserText;
+  if (!value || typeof value !== 'object' || Array.isArray(value)) return null;
+  const answer = value as Record<string, unknown>;
+  return typeof answer.text === 'string' && typeof answer.acceptedAt === 'number'
+    && Number.isFinite(answer.acceptedAt)
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/maker-ipc/autoReviewUserIntent.ts:21]: Fixed in 73bb4c931. The shared messages.jsonl parser strips autoReviewUserText before persistence for every role, harness, and Orca worker, while preserving other metadata and Pi transcript rewriting. Real bundle import tests cover forged user-string/card-object grants across Claude, Codex, Pi and a worker; all 50 import tests, related unit gate, and Desktop typecheck pass. Independently re-reviewed the import boundary; native transcript importers construct metadata from explicit fields.

File: apps/desktop/src/main/maker-ipc/agent-input-coordinator.ts [hunk 2]
```diff
@@ -2140,6 +2144,9 @@ export class AgentInputCoordinator {
         referenceContexts,
       );
       await this.deps.steerToAgent(sessionId, buildMakerUserMessage(item, referenceContexts), {
+        [AUTO_REVIEW_SOURCE_CONTENT]: item.autoReviewUserText ?? '',
+        ...(readAutoReviewUserText(item.persistedContent) === null
+          ? { [AUTO_REVIEW_USER_INTENT]: item.autoReviewUserText ?? '' } : {}),
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/maker-ipc/agent-input-coordinator.ts:2149]: Fixed in a95fc0d0d at the shared Host steer entry, covering queued and direct steers. It reuses the send-path persistence barrier and bounded authorization query, preserving accepted clarification/plan receipts. Resource replacements remain explicit, external identities cannot inherit owner history, and cancellation plus Session/turn/input-boundary checks run after history loading. Regression coverage includes reconnect-style follow-ups, plugin rewrites, attachments, unavailable/cleared history and cancellation. Related unit gate, Desktop typecheck and independent review passed.

**Review by MagicLizi — APPROVED** [review]
## PR #4078 Code Review Report

**Title**: fix(auto-review): 保留真实授权与澄清限制，准确回传拒绝原因
**Author**: dashhuang (maintainer)
**Review tier**: standard
**Diff**: 1246 lines, 37 files
**Verdict**: Clean — 0 P0, 0 P1

### Summary

This PR fixes three related issues in the Auto permission review system:

1. **Authorization capture at Host boundary**: User's original text is now captured before plugin rewrites, session decoration, and handoff injection. Stored independently in `agentMeta.autoReviewUserText` via atomic writes.

2. **Authorization history restoration**: New `autoReviewUserIntent.ts` module restores bounded authorization history from persisted messages, including clarification answers and plan review results, ordered by acceptance time. Handles clear/rewind boundaries, oversized contexts, timestamp ties, and plugin rewrites.

3. **Denial reason propagation**: Specific denial reasons (auto-review block, user denial, system cancellation) are now propagated through all three harnesses (Claude Code, Codex, Pi) via `formatPermissionDenial`. Previously, all denials showed generic messages.

### Security Review

**IPC boundary protections verified:**
- Renderer/caller cannot mint, replace, or remove Host-authored `autoReviewUserText` via any IPC path (create, updateAgentMeta, updateContent)
- Session share imports strip `autoReviewUserText` from all messages across all roles and Orca workers
- Non-desktop origins (IM, device-link) do not inherit owner authorization history
- Content edits by users invalidate prior authorization (user role only); card PATCHes cannot alter Host-accepted answers
- Plugin rewrites detected: original text preserved in separate metadata field
- Quoted text, session references, pasted ranges, and resource attachments correctly invalidate plain-text authorization
- Oversized history drops all prior context rather than sampling (preserves revocations)
- Timestamp ties between cards and messages safely return empty authorization

**Auto-review prompt changes verified:**
- Restructured decision order (AUTHORITY → SCOPE → VERDICT) with clearer semantics
- `writableRoots` renamed to `defaultWritableRoots` with clarified non-restrictive semantics
- Always provides `authorizationContext` (defaults to owner/direct)
- System/user channel separation for model requests (rules on system, evidence on user)
- Model output reasons redacted via `redactSensitiveText`

**Codex approval flow changes verified:**
- `resolveApproval` return type changed from `ApprovalDecision` to `{decision, reason?}`
- All 10+ call sites updated consistently
- `settled` guard prevents late UI replies from overwriting cancellation reasons
- Pi package management changed from `cancelled:true` to structured error JSON

### Test Coverage

- New `autoReviewUserIntent.test.ts` (203 lines): authorization restoration, ordering, plugin detection, resource invalidation, oversized handling
- New `makerSendTransaction.test.ts` additions (81 lines): capture, quotes, images, history restoration, oversized input
- New `messagesListCursor.test.ts` additions (89 lines): IPC forgery prevention, atomic writes, authorization query, clear/rewind
- Updated harness tests across Claude Code, Codex, Pi for denial reason format
- Session share import stripping tests for all roles and Orca workers
- Ghost tool cancellation reason distinction tests

### CI

- client-ci: pass
- pr-design-basis: pass
- All status checks green
